pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,64 +15,240 @@
+		"sourceNodeName",
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+		"destinationClusterIPv4",
+		"destinationServicePortName",
+	}
+	aggregationFields = []string{
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+	}
 )
 
 type flowAggregator struct {
-	flowCollectorAddr net.Addr
-	exportFrequency   time.Duration
+	flowCollectorAddr  net.Addr
+	collectingProcess  ipfix.IPFIXCollectingProcess
+	aggregationProcess ipfix.IPFIXAggregationProcess
+	exportFrequency    time.Duration
+	exportingProcess   ipfix.IPFIXExportingProcess
+	templateID         uint16
+	registry           ipfix.IPFIXRegistry
+	sentFlowKeys       []*ipfixintermediate.FlowKey
 }
 
-func InitFlowAggregator(flowCollectorAddr net.Addr, exportFrequency time.Duration) *flowAggregator {
+func NewFlowAggregator(flowCollectorAddr net.Addr, exportFrequency time.Duration) *flowAggregator {
+	registry := ipfix.NewIPFIXRegistry()
 	registry.LoadRegistry()
 	return &flowAggregator{
 		flowCollectorAddr,
+		nil,
+		nil,
 		exportFrequency,
+		nil,
+		0,
+		registry,
+		make([]*ipfixintermediate.FlowKey, 0),
+	}
+}
+
+func genObservationID() (uint32, error) {
+	name, err := env.GetNodeName()
+	if err != nil {
+		return 0, err
+	}
+	h := fnv.New32()
+	h.Write([]byte(name))
+	return h.Sum32(), nil
+}
+
+func (fag *flowAggregator) initExportingProcess() error {
+	obsID, err := genObservationID()
+	if err != nil {
+		return fmt.Errorf("cannot generate obsID for IPFIX ipfixexport: %v", err)
+	}
+	var ep ipfix.IPFIXExportingProcess
+	if fag.flowCollectorAddr.Network() == "tcp" {
+		// TCP transport do not need any tempRefTimeout, so sending 0.
+		ep, err = ipfix.NewIPFIXExportingProcess(fag.flowCollectorAddr, obsID, 0)
+	} else {
+		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
+		ep, err = ipfix.NewIPFIXExportingProcess(fag.flowCollectorAddr, obsID, 1800)
+	}
+	if err != nil {
+		return fmt.Errorf("Got error when initializing IPFIX exporting process: %v", err)
+	}
+	fag.exportingProcess = ep
+	fag.templateID = fag.exportingProcess.NewTemplateID()
+	templateSet := ipfix.NewSet(ipfixentities.Template, fag.templateID, false)
+
+	bytesSent, err := fag.sendTemplateRecord(templateSet)
+	if err != nil {
+		fag.exportingProcess = nil
+		fag.templateID = 0
+		return fmt.Errorf("Send template message failed, err: %v", err)
 	}
+	klog.V(2).Infof("Initialized exporting process and sent %d bytes size of template record", bytesSent)
+	return nil
 }
 
-func (fae *flowAggregator) Run(stopCh <-chan struct{}) {
+func (fag *flowAggregator) Run(stopCh <-chan struct{}) {
 	// Go-ipfix TCP collector doesn't work during my last test, fixed as udp temporarily.
 	collectAddr, _ := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
-	cp, _ := collector.InitCollectingProcess(collectAddr, 1024, 0)
 
-	ep, err := exporter.InitExportingProcess(fae.flowCollectorAddr, 1, 0)
+	cp, err := ipfix.NewIPFIXCollectingProcess(collectAddr, 1024, 0)
```

Comment:
TCP may not be working because of maxBufferSize. Please try 65535 as the arg--it may work. Even though sent TCP packets are small, segmentation offload may come into the picture.

Reply:
Thanks, TCP collecting process works well after try 65535 maxBufferSize.